File: lib/scripts/common/coalheap.php
<?php
/*
class CoalHeap for cqm scripts
Copyright (c) Ruslan Keba 2004
*/

define("P_MIN", 0.1);
define("N_MIN", 0.1);
define("A_MIN", 2.0);
define("A_MAX", 80.0);
define("H_MIN", 2.0);
define("H_MAX", 80.0);

class Coal    //уголь
{
    function Coal($p=0, $a=0, $h=0)//конструктор
    {
        $this->P = $p;        
        $this->A = $a;        
        $this->H = $h;        
    }
    
    function cempty()//true если p=a=h=0;
    {
        if ($this->P == 0 && $this->A == 0 && $this->H == 0)
        {
            return TRUE;
        }
        else
        {
            return FALSE;
        }
    }

    function setempty()//set p=a=h=0;
    {
        $this->P = 0;
        $this->A = 0;
        $this->H = 0;
    }

    var $P;//вес угля
    var $A;//зольность угля
    var $H;//влажность угля

    function format()
    {
        return "[".$this->P.",".$this->A.",".$this->c->H."]";
    }
}

class CoalHeap
{
    var $c;  //уголь
    var $n;  //количесво циклов за который добавляли или отнимали уголь
    var $dn;
    var $empty; //пустота кучи

    function CoalHeap()
    {
        $this->c = new Coal;
        $this->setEmpty();
        $this->resetDeltaN();
    }

    function addCoal ($m)//добавить угля
    {
        if ($m->cempty())
            return;

        $this->addAsh($m);
        $this->addHum($m);
        $this->addWei($m);
        $this->n += $dn;
        $this->correctCoal();
        $this->resetDeltaN();
    }

    function subCoal ($m)//отнять угля
    {
        if ($m->cempty())
            return;

        $this->subAsh($m);
        $this->subHum($m);
        $this->subWei($m);
        $this->$n -= $dn;
        $this->correctCoal();
        $this->resetDeltaN();
    }

    function addAsh($m)
    {
        if (is_null($m->A) || $m->A<=A_MIN)
            return;

        if (is_null($this->c->A) || $this->c->A<=A_MIN)
            $this->c->A = $m->A;

        if (!is_null($this->c->P) && !is_null($m->P) && ($this->c->P + $m->P)>=P_MIN)
        {
             $this->c->A = ($this->c->A * $this->c->P + $m->A * $m->P)/($this->c->P + $m->P);
             return;
        }

        $this->c->A = ($this->c->A * $this->n + $m->A->val * $this->dn)/($this->n + $this->dn);
    }

    function addHum($m)
    {
        if (is_null($m->H) || $m->H<=H_MIN)
            return;

        if (is_null($this->c->H) || $this->c->H<=H_MIN)
            $this->c->H = $m->H;

        if (!is_null($this->c->P) && !is_null($m->P) && ($this->c->P + $m->P)>=P_MIN)
        {
             $this->c->H = ($this->c->H * $this->c->P + $m->H * $m->P)/($this->c->P + $m->P);
             return;
        }

        $this->c->H = ($this->c->H * $this->n + $m->H->val * $this->dn)/($this->n + $this->dn);
    }

    function addWei($m)
    {
        if (is_null($m->P) || $m->P<=P_MIN)
            return;

        if (is_null($this->c->P))
            return;

        $this->c->P += $m->P;
    }

    function subAsh($m)
    {
        if (is_null($m->A) || $m->A<=A_MIN)
            return;

        if (is_null($this->c->A) || $this->c->A<=A_MIN)
            $this->c->A = $m->A;

        if (!is_null($this->c->P) && !is_null($m->P) && ($this->c->P - $m->P)>=P_MIN)
        {
             $this->c->A = ($this->c->A * $this->c->P - $m->A * $m->P)/($this->c->P - $m->P);
             return;
        }

        if ($this->n - $this->dn > N_MIN )
            $this->c->A = ($this->c->A * $this->n - $m->A * $this->dn)/($this->n - $this->dn);
    }

    function subHum($m)
    {
        if (is_null($m->H) || $m->H<=H_MIN)
            return;

        if (is_null($this->c->H) || $this->c->H<=H_MIN)
            $this->c->H = $m->H;

        if (!is_null($this->c->P) && !is_null($m->P) && ($this->c->P - $m->P)>=P_MIN)
        {
             $this->c->H = ($this->c->H * $this->c->P - $m->H * $m->P)/($this->c->P - $m->P);
             return;
        }

        if ($this->n - $this->dn > N_MIN )
            $this->c->H = ($this->c->H * $this->n - $m->H * $this->dn)/($this->n - $this->dn);
    }

    function subWei($m)
    {
        if (is_null($m->P) || $m->P<=P_MIN)
            return;

        if (is_null($this->c->P))
            return;

        $this->c->P -= $m->P;
    }

    function correctCoal()
    {
        if ($this->c->A <= A_MIN || $this->c->A >= A_MAX)
            $this->c->A = NULL;

        if ($this->c->H <= H_MIN || $this->c->H >= H_MAX)
            $this->c->H = NULL;

        if ($this->c->P <= P_MIN)
            $this->setEmpty();
        else
            $this->setNotEmpty();

        if ($this->n <= N_MIN)
            $this->n = 0;
    }

    function setEmpty()//опустошить
    {
        $this->c->setempty();
        $this->n = 0;
        $this->empty = 1;
    }

    function setNotEmpty()//не опустошить
    {
        $this->empty = 0;
    }

    function setDeltaN($dn)//установить dn
    {
        $this->dn = $dn;
        if ($this->dn <= 0)
            $this->resetDeltaN();
    }


    function resetDeltaN()//сбросить dn
    {
        $this->dn = 1.0;
    }

    function format()
    {
        return "(".$this->c->format().",".$this->n.",".$this->empty.")";
    }
}

/* tests */

/*
$ch = new CoalHeap;
$ch->c->P = 10;
$ch->c->A = 30;
print "(".$ch->c->P.",".$ch->c->A.")\n";

$c = new Coal(10, 20, 0);
print "(".$c->P.",".$c->A.")\n";

$ch->addCoal($c);
print "(".$ch->c->P.",".$ch->c->A.")\n";
$ch->subCoal($c);
print "(".$ch->c->P.",".$ch->c->A.")\n";
*/

?>

File: common/classes/coalheap.cpp
//---------------------------------------------------------------------------
#include "coalheap.h"
#include "defines.h"

//---------------------------------------------------------------------------
CoalHeap::CoalHeap(void)
{
    setEmpty();
    resetDeltaN();
}

//---------------------------------------------------------------------------
void CoalHeap::setEmpty(void)
{
    c = Coal(0, 0);
    Pa.val = 0;
    n  = 0;
    empty = true;
}

//---------------------------------------------------------------------------
void CoalHeap::setNotEmpty(void)
{
    empty = false;
}

//---------------------------------------------------------------------------
void CoalHeap::setPa()
{
    Pa = c.P;
}

//---------------------------------------------------------------------------
void CoalHeap::addCoal (Coal m)
{
    if (m.empty())
        return;

    addAsh(m);
    addHum(m);
    addWei(m);
    n += dn;
    correctCoal();
    resetDeltaN();
}

//---------------------------------------------------------------------------
void CoalHeap::addAsh (Coal m)
{
/*
    if (m.A.isNull || m.A.val<=A_MIN)
        return;

    if (c.A.isNull || c.A.val<=A_MIN)
        c.A = m.A;

    if (!c.P.isNull && !m.P.isNull && (c.P.val + m.P.val)>=P_MIN)
    {
         c.A.val = (c.A.val * c.P.val + m.A.val * m.P.val)/(c.P.val + m.P.val);
         return;
    }

    c.A.val = (c.A.val * n + m.A.val * dn)/(n + dn);
*/
    if (m.A.isNull || m.A.val<=A_MIN
     || m.P.isNull || m.P.val<=P_MIN)
        return;

    if (Pa.isNull || Pa.val<=P_MIN)
    {
        c.A = m.A;
        Pa  = m.P;
        return;
    }

    c.A.val = (c.A.val * Pa.val + m.A.val * m.P.val)/(Pa.val + m.P.val);
    Pa.val += m.P.val;
}

//---------------------------------------------------------------------------
//по идее нужно сделать так же как и для ash
void CoalHeap::addHum (Coal m)
{
    if (m.H.isNull || m.H.val<=H_MIN)
        return;

    if (c.H.isNull || c.H.val<=H_MIN)
        c.H = m.H;

    if (!c.P.isNull && !m.P.isNull && (c.P.val + m.P.val)>=P_MIN)
    {
         c.H.val = (c.H.val * c.P.val + m.H.val * m.P.val)/(c.P.val + m.P.val);
         return;
    }

    c.H.val = (c.H.val * n + m.H.val * dn)/(n + dn);
}

//---------------------------------------------------------------------------
void CoalHeap::addWei (Coal m)
{
    if (m.P.isNull || m.P.val<=P_MIN)
        return;

    if (c.P.isNull)
        return;

    c.P.val += m.P.val;
}

//---------------------------------------------------------------------------
void CoalHeap::subCoal (Coal m)
{
    if (m.empty())
        return;

    subAsh(m);
    subHum(m);
    subWei(m);
    n -= dn;
    correctCoal();
    resetDeltaN();
}

//---------------------------------------------------------------------------
void CoalHeap::subAsh (Coal m)
{
/*
    if (m.A.isNull || m.A.val<=A_MIN)
        return;

    if (c.A.isNull || c.A.val<=A_MIN)
        c.A = m.A; // можно ли так?

    if (!c.P.isNull && !m.P.isNull && (c.P.val - m.P.val)>=P_MIN)
    {
         c.A.val = (c.A.val * c.P.val - m.A.val * m.P.val)/(c.P.val - m.P.val);
         return;
    }

    if (n - dn > N_MIN )
        c.A.val = (c.A.val * n - m.A.val * dn)/(n - dn);
*/
    if (m.A.isNull || m.A.val<=A_MIN
     || m.P.isNull || m.P.val<=P_MIN)
        return;

    if (Pa.isNull || Pa.val<=A_MIN)
    {
        return;
    }

    if ((Pa.val - m.P.val) >= P_MIN)
    {
         c.A.val = (c.A.val * Pa.val - m.A.val * m.P.val)/(Pa.val - m.P.val);
         Pa.val -= m.P.val;
         return;
    }
    else
    {
        Pa.val = 0;
    }
}

//---------------------------------------------------------------------------
void CoalHeap::subHum (Coal m)
{
    if (m.H.isNull || m.H.val<=H_MIN)
        return;

    if (c.H.isNull || c.H.val<=H_MIN)
        c.H = m.H; // можно ли так?

    if (!c.P.isNull && !m.P.isNull && (c.P.val - m.P.val)>P_MIN)
    {
         c.H.val = (c.H.val * c.P.val - m.H.val * m.P.val)/(c.P.val - m.P.val);
         return;
    }

    if (n - dn > N_MIN )
        c.H.val = (c.H.val * n - m.H.val * dn)/(n - dn);
}

//---------------------------------------------------------------------------
void CoalHeap::subWei (Coal m)
{
    if (m.P.isNull || m.P.val<=P_MIN)
        return;

    if (c.P.isNull)
        return;

    c.P.val -= m.P.val;
}

//---------------------------------------------------------------------------
void CoalHeap::correctCoal(void)
{
    if (c.A.val <= A_MIN || c.A.val >= A_MAX)
        c.A.setNull();

    if (c.H.val <= H_MIN || c.H.val >= H_MAX)
        c.H.setNull();

    if (c.P.val <= P_MIN)
        setEmpty();
    else
        setNotEmpty();

    if (n <= N_MIN)
        n = 0;

}

//---------------------------------------------------------------------------
void CoalHeap::setDeltaN(double dN_)
{
    dn = dN_;
    if (dn <= 0) resetDeltaN();
}

//---------------------------------------------------------------------------
void CoalHeap::resetDeltaN(void)
{
    dn = 1.0;
}


//---------------------------------------------------------------------------
ostream& operator<<(ostream& s, CoalHeap& ch)
{
    return s<<ch.c<<" "<<ch.n<<" "<<ch.empty;
}

//---------------------------------------------------------------------------
istream& operator>>(istream& s, CoalHeap& ch)
{
    return s>>ch.c>>ch.n>>ch.empty;
}

//---------------------------------------------------------------------------
double HeatVal(Coal c, double hhv, double k)
{
    double ret = hhv*(1-c.A.val/100.0)*(1-c.H.val/100.0) -
                 2.442*(c.H.val/100.0+k*(1-c.A.val/100.0)*(1-c.H.val/100.0));
    return ret;
}

//---------------------------------------------------------------------------
AverageWeightedHeat::AverageWeightedHeat(void)
{
    setEmpty();
}

//---------------------------------------------------------------------------
void AverageWeightedHeat::add(Coal c, double hv)
{
    if (p+c.P.val > P_MIN)
    {
        heat = (heat*p + c.P.val*hv) / (p += c.P.val);
    }
}

//---------------------------------------------------------------------------
void AverageWeightedHeat::setEmpty(void)
{
    p = 0;
    heat = 0;
}

//---------------------------------------------------------------------------
double AverageWeightedHeat::getHeat(void)
{
    return heat;
}

//---------------------------------------------------------------------------

